File: frontend/src/data/buildPlans.ts
/** Build packages shown after a live preview — choose plan, then reach out. */

export interface BuildAddon {
  id: string;
  name: string;
  description: string;
  /** Soft floor in USD; exact quote confirmed manually. */
  fromUsd: number;
  /** Plans that already include this addon (cannot toggle off as extra). */
  includedIn?: Array<'launch' | 'growth'>;
  /** Why we suggested this for this business (shown as a small hint). */
  whyForYou?: string;
}

export interface BuildPlan {
  id: 'launch' | 'growth' | 'custom';
  name: string;
  tagline: string;
  /** Soft floor; null = custom quote only */
  fromUsd: number | null;
  timeline: string;
  highlight?: boolean;
  badge?: string;
  includes: string[];
  bestFor: string;
}

export interface BuildAddonContext {
  businessName?: string | null;
  conceptName?: string | null;
  industry?: string | null;
  mainProblem?: string | null;
  desiredOutcome?: string | null;
  previewFeatures?: string[];
  aiFeatures?: Array<{ id?: string; name?: string; description?: string }>;
  roleLabels?: string[];
}

export const BUILD_PLANS: BuildPlan[] = [
  {
    id: 'launch',
    name: 'Launch MVP',
    tagline: 'Ship the preview you just saw — for real.',
    fromUsd: 3500,
    timeline: '4–8 weeks',
    bestFor: 'Getting live fast with your core customer + owner flows',
    includes: [
      'Production build of your preview (public site + core journeys)',
      'Owner / admin basics',
      '1–2 AI features from your plan',
      'Brand colors, typography, and deploy',
      'Launch handoff + walkthrough',
    ],
  },
  {
    id: 'growth',
    name: 'Growth MVP',
    tagline: 'Launch + make it earn and run itself.',
    fromUsd: 7500,
    timeline: '8–12 weeks',
    highlight: true,
    badge: 'Most popular',
    bestFor: 'Teams that want payments, AI, and polish in one go',
    includes: [
      'Everything in Launch MVP',
      'Payments / booking / ordering polish',
      'Extra AI features + automations',
      'WhatsApp or SMS reminders (where applicable)',
      'More pages & roles as scoped',
      '30 days of post-launch tweaks',
    ],
  },
  {
    id: 'custom',
    name: 'Custom / Scale',
    tagline: 'Integrations, multi-location, or ongoing product.',
    fromUsd: null,
    timeline: 'Scoped together',
    bestFor: 'Complex ops, POS/CRM links, or a longer partnership',
    includes: [
      'Everything you need beyond Growth',
      'Third-party integrations',
      'Advanced roles & permissions',
      'Custom workflows and reporting',
      'Dedicated scope call before we quote',
    ],
  },
];

type IndustryBucket =
  | 'food'
  | 'beauty'
  | 'fitness'
  | 'health'
  | 'retail'
  | 'hospitality'
  | 'education'
  | 'services'
  | 'generic';

function bucketIndustry(industry: string | null | undefined): IndustryBucket {
  const t = (industry || '').toLowerCase();
  if (/restaurant|cafe|coffee|food|bakery|bistro|dining|bar/.test(t)) return 'food';
  if (/spa|salon|beauty|skin|aesthetic|wellness|massage|nail/.test(t)) return 'beauty';
  if (/fitness|gym|yoga|pilates|coach|training/.test(t)) return 'fitness';
  if (/clinic|dental|health|medical|therapy|physio/.test(t)) return 'health';
  if (/retail|fashion|shop|boutique|store|ecommerce/.test(t)) return 'retail';
  if (/hotel|hospitality|inn|resort/.test(t)) return 'hospitality';
  if (/school|tutor|education|course|academy|learning/.test(t)) return 'education';
  if (/service|trade|plumb|hvac|clean|agency|law|legal/.test(t)) return 'services';
  return 'generic';
}

function blob(ctx: BuildAddonContext): string {
  return [
    ctx.industry,
    ctx.mainProblem,
    ctx.desiredOutcome,
    ...(ctx.previewFeatures || []),
    ...(ctx.aiFeatures || []).map((f) => `${f.name || ''} ${f.description || ''}`),
    ...(ctx.roleLabels || []),
  ]
    .filter(Boolean)
    .join(' ')
    .toLowerCase();
}

/**
 * Suggest add-ons tailored to this business / preview — not a generic catalog.
 * Always returns a short list (≈5–7) with clear “why for you” hints.
 */
export function suggestBusinessAddons(ctx: BuildAddonContext): BuildAddon[] {
  const brand = (ctx.businessName || ctx.conceptName || 'your business').trim();
  const bucket = bucketIndustry(ctx.industry);
  const text = blob(ctx);
  const aiNames = (ctx.aiFeatures || [])
    .map((f) => (f.name || '').trim())
    .filter(Boolean)
    .slice(0, 3);
  const roles = (ctx.roleLabels || []).map((r) => r.trim()).filter(Boolean);
  const hasBooking = /book|reserv|appoint|table|slot/.test(text);
  const hasOrder = /order|menu|cart|checkout|pickup|delivery/.test(text);
  const hasNoShow = /no-?show|phone|remind|whatsapp|sms/.test(text);
  const hasInventory = /inventory|stock|sku|catalog/.test(text);
  const out: BuildAddon[] = [];

  const push = (addon: BuildAddon) => {
    if (out.some((a) => a.id === addon.id)) return;
    out.push(addon);
  };

  // 1) AI — name real features from their preview when present
  if (aiNames.length >= 2) {
    push({
      id: 'ai-pack',
      name: `Full AI suite for ${brand}`,
      description: `Wire all previewed AI: ${aiNames.join(', ')}.`,
      fromUsd: 1200,
      includedIn: ['growth'],
      whyForYou: 'These AI features are already in your preview plan.',
    });
  } else if (aiNames.length === 1) {
    push({
      id: 'ai-pack',
      name: `Production-ready ${aiNames[0]}`,
      description: `Take “${aiNames[0]}” from demo to live with your real data and guardrails.`,
      fromUsd: 900,
      includedIn: ['growth'],
      whyForYou: 'Highlighted in your AI hub.',
    });
  } else {
    push({
      id: 'ai-pack',
      name: `AI assistant for ${brand}`,
      description: 'A branded AI helper trained on your offerings and FAQs.',
      fromUsd: 800,
      includedIn: ['growth'],
      whyForYou: 'Speeds up answers your customers already ask.',
    });
  }

  // 2) Industry-primary commerce / ops addon
  if (bucket === 'food' || hasOrder) {
    push({
      id: 'payments',
      name: 'Online ordering + payments',
      description: `Card checkout for ${brand} pickup/orders — cart through paid confirmation.`,
      fromUsd: 1400,
      includedIn: ['growth'],
      whyForYou: 'Cuts phone orders and missed tickets.',
    });
  } else if (bucket === 'beauty' || bucket === 'health' || bucket === 'fitness' || hasBooking) {
    push({
      id: 'payments',
      name: 'Booking deposits & payments',
      description: 'Take deposits or full payment when customers book — fewer no-shows.',
      fromUsd: 1200,
      includedIn: ['growth'],
      whyForYou: hasNoShow
        ? 'Your brief called out no-shows / phone friction.'
        : 'Booking businesses convert better with deposits.',
    });
  } else if (bucket === 'retail') {
    push({
      id: 'payments',
      name: 'Shop checkout + catalog',
      description: 'Sell featured products with real checkout and order status.',
      fromUsd: 1300,
      includedIn: ['growth'],
      whyForYou: 'Turns your showcase into revenue.',
    });
  } else if (bucket === 'hospitality') {
    push({
      id: 'payments',
      name: 'Stay / experience booking payments',
      description: 'Secure reservations with deposits and confirmation emails.',
      fromUsd: 1300,
      includedIn: ['growth'],
      whyForYou: 'Hospitality needs paid holds, not open calendars.',
    });
  } else {
    push({
      id: 'payments',
      name: 'Payments for your main journey',
      description: 'Checkout wired into the primary action in your preview.',
      fromUsd: 1200,
      includedIn: ['growth'],
      whyForYou: 'Makes the demo path collect real money.',
    });
  }

  // 3) Messaging / reminders — industry voice
  if (bucket === 'food') {
    push({
      id: 'messaging',
      name: 'Order & table reminders',
      description: 'WhatsApp/SMS for order-ready alerts and reservation reminders.',
      fromUsd: 900,
      includedIn: ['growth'],
      whyForYou: hasNoShow
        ? 'Matches your no-show / phone-order pain.'
        : 'Keeps tables and pickups on time.',
    });
  } else if (bucket === 'beauty' || bucket === 'health' || bucket === 'fitness') {
    push({
      id: 'messaging',
      name: 'Appointment reminder automations',
      description: 'WhatsApp/SMS before visits + easy reschedule links.',
      fromUsd: 900,
      includedIn: ['growth'],
      whyForYou: 'Protects your calendar from empty slots.',
    });
  } else {
    push({
      id: 'messaging',
      name: 'Customer messaging automations',
      description: 'Confirmations and follow-ups on WhatsApp or SMS.',
      fromUsd: 900,
      includedIn: ['growth'],
      whyForYou: 'Where your customers already reply.',
    });
  }

  // 4) Roles — use real role labels when we have them
  if (roles.length >= 2) {
    push({
      id: 'roles',
      name: `Dashboards for ${roles.slice(0, 3).join(' + ')}`,
      description: `Separate live workspaces for each role in your preview.`,
      fromUsd: 800,
      whyForYou: 'Your preview already defines these roles.',
    });
  } else if (bucket === 'food') {
    push({
      id: 'roles',
      name: 'Floor + kitchen ops views',
      description: 'Staff queue / ticket board separate from the owner dashboard.',
      fromUsd: 750,
      whyForYou: 'Cafe ops need more than one admin screen.',
    });
  } else if (bucket === 'beauty' || bucket === 'fitness') {
    push({
      id: 'roles',
      name: 'Front desk + practitioner views',
      description: 'Reception books; practitioners see today’s clients.',
      fromUsd: 750,
      whyForYou: 'Two jobs, two screens — fewer mix-ups.',
    });
  } else {
    push({
      id: 'roles',
      name: 'Extra staff / partner roles',
      description: 'Additional dashboards beyond the owner view.',
      fromUsd: 700,
      whyForYou: 'When more than one person runs the product daily.',
    });
  }

  // 5) Industry specialty
  if (bucket === 'food') {
    push({
      id: 'specialty',
      name: 'Live menu & allergen data',
      description: 'Editable menu, modifiers, and allergen facts feeding your AI answers.',
      fromUsd: 850,
      whyForYou: /allergen|diet/.test(text)
        ? 'Your plan includes allergen / dietary AI.'
        : 'Menus change — the app should too.',
    });
  } else if (bucket === 'beauty' || bucket === 'health') {
    push({
      id: 'specialty',
      name: 'Client notes & rebooking',
      description: 'Treatment history, photos, and smart rebook prompts.',
      fromUsd: 850,
      whyForYou: 'Repeat visits are the business model.',
    });
  } else if (bucket === 'fitness') {
    push({
      id: 'specialty',
      name: 'Class packs & memberships',
      description: 'Sell packs, track credits, and gate class booking.',
      fromUsd: 950,
      whyForYou: 'Studios grow on packs, not one-off visits.',
    });
  } else if (bucket === 'retail' || hasInventory) {
    push({
      id: 'specialty',
      name: 'Inventory & low-stock alerts',
      description: 'Track SKUs and warn before bestsellers run out.',
      fromUsd: 800,
      whyForYou: 'Retail dies when the shelf is empty online.',
    });
  } else if (bucket === 'education') {
    push({
      id: 'specialty',
      name: 'Class schedule & enrollments',
      description: 'Sessions, seats, and student confirmations.',
      fromUsd: 850,
      whyForYou: 'Education needs seats, not a static brochure.',
    });
  } else if (bucket === 'services') {
    push({
      id: 'specialty',
      name: 'Job / lead pipeline',
      description: 'Capture leads, assign jobs, and track status to done.',
      fromUsd: 900,
      whyForYou: 'Service businesses win on follow-through.',
    });
  } else {
    push({
      id: 'specialty',
      name: `Domain module for ${brand}`,
      description: 'A custom workflow module matched to your industry preview.',
      fromUsd: 850,
      whyForYou: 'Goes beyond a generic website template.',
    });
  }

  // 6) Always useful polish + care (shorter why)
  push({
    id: 'cinema',
    name: `Cinematic polish for ${brand}`,
    description: 'Hero photography art-direction, motion, and brand-first craft.',
    fromUsd: 600,
    whyForYou: 'Makes the live product feel as premium as the preview.',
  });

  push({
    id: 'care',
    name: '30-day care after launch',
    description: 'Priority fixes and small iterations once you’re live.',
    fromUsd: 500,
    includedIn: ['growth'],
    whyForYou: 'First month is when real customers find edge cases.',
  });

  return out.slice(0, 7);
}

/** Fallback catalog if context is empty (landing page, etc.). */
export const BUILD_ADDONS: BuildAddon[] = suggestBusinessAddons({
  businessName: 'your business',
  industry: 'general',
});

export function formatFromUsd(amount: number | null): string {
  if (amount == null) return 'Custom quote';
  return `From $${amount.toLocaleString('en-US')}`;
}

export function planBaseFrom(planId: BuildPlan['id']): number | null {
  return BUILD_PLANS.find((p) => p.id === planId)?.fromUsd ?? null;
}

export function addonAvailable(addon: BuildAddon, planId: BuildPlan['id']): boolean {
  if (planId === 'custom') return true;
  return !(addon.includedIn || []).includes(planId as 'launch' | 'growth');
}

export function addonIncluded(addon: BuildAddon, planId: BuildPlan['id']): boolean {
  if (planId === 'custom') return false;
  return (addon.includedIn || []).includes(planId as 'launch' | 'growth');
}

/** Soft estimate: plan floor + selected add-ons (custom stays null). */
export function estimateFromUsd(
  planId: BuildPlan['id'],
  selectedAddonIds: string[],
  catalog: BuildAddon[] = BUILD_ADDONS,
): number | null {
  const base = planBaseFrom(planId);
  if (base == null) return null;
  const extra = catalog
    .filter((a) => selectedAddonIds.includes(a.id) && addonAvailable(a, planId))
    .reduce((sum, a) => sum + a.fromUsd, 0);
  return base + extra;
}

export function summarizeSelection(
  planId: BuildPlan['id'],
  selectedAddonIds: string[],
  catalog: BuildAddon[] = BUILD_ADDONS,
): string {
  const plan = BUILD_PLANS.find((p) => p.id === planId);
  const addons = catalog.filter(
    (a) => selectedAddonIds.includes(a.id) && addonAvailable(a, planId),
  );
  const estimate = estimateFromUsd(planId, selectedAddonIds, catalog);
  const parts = [
    `Package: ${plan?.name || planId}`,
    estimate != null
      ? `Estimate: ${formatFromUsd(estimate)} USD (soft floor — exact quote after scope)`
      : 'Estimate: custom quote after scope call',
  ];
  if (addons.length) {
    parts.push(
      `Add-ons: ${addons.map((a) => `${a.name} (+$${a.fromUsd})`).join('; ')}`,
    );
  } else {
    parts.push('Add-ons: none');
  }
  return parts.join('\n');
}

File: backend/app/application/pipelines/build_plans.py
"""Generate structured Launch/Growth/Custom build plans + add-ons (no prices)."""
from __future__ import annotations

import json
import re
from datetime import datetime
from typing import Any

from sqlalchemy.orm import Session

from app.application.prompts import PromptTemplate
from app.application.pipelines._shared import get_request
from app.application.services.ai_features import ai_features_from_request
from app.core.config import settings
from app.domain.interfaces.ai_provider import AIProvider
from app.domain.interfaces.template_renderer import TemplateRenderer
from app.shared.json_utils import extract_json_from_text

_PLAN_IDS = ("launch", "growth", "custom")
_MONEY_RE = re.compile(
    r"(\$\s*\d|\bUSD\b|\bEUR\b|\bprice\b|\bcost\b|\bfrom\s+\$|\bquote\s+of\b)",
    re.I,
)


def _role_labels(req) -> list[str]:
    if not req.generated_pages:
        return []
    try:
        bundle = json.loads(req.generated_pages)
    except Exception:
        return []
    roles = (bundle.get("preview_app") or {}).get("roles") or bundle.get("roles") or []
    out: list[str] = []
    for r in roles:
        if isinstance(r, dict):
            label = (r.get("label") or r.get("id") or "").strip()
            if label:
                out.append(label)
        elif isinstance(r, str) and r.strip():
            out.append(r.strip())
    return out


def _preview_features(req) -> list[str]:
    if not req.preview_features:
        return []
    try:
        data = json.loads(req.preview_features)
        if isinstance(data, list):
            return [str(x) for x in data if str(x).strip()]
    except Exception:
        pass
    return [line.strip() for line in str(req.preview_features).splitlines() if line.strip()]


def _scrub_money(text: str) -> str:
    if not text:
        return text
    if not _MONEY_RE.search(text):
        return text
    # Soft scrub — drop obvious price fragments
    cleaned = re.sub(r"\$\s*[\d,]+(?:\.\d+)?", "", text)
    cleaned = re.sub(r"\bfrom\s+\d[\d,]*(?:\.\d+)?\s*(USD|usd)?\b", "", cleaned, flags=re.I)
    return re.sub(r"\s{2,}", " ", cleaned).strip()


def _normalize_plan(raw: dict[str, Any], plan_id: str) -> dict[str, Any]:
    includes = raw.get("includes") or []
    if not isinstance(includes, list):
        includes = [str(includes)]
    includes = [_scrub_money(str(x)) for x in includes if str(x).strip()][:10]
    if not includes:
        includes = ["Production build of your preview", "Owner / admin basics", "Launch handoff"]
    badge = raw.get("badge")
    if badge is not None:
        badge = _scrub_money(str(badge)) or None
    return {
        "id": plan_id,
        "name": _scrub_money(str(raw.get("name") or plan_id.title()))[:80],
        "tagline": _scrub_money(str(raw.get("tagline") or ""))[:200],
        "timeline": _scrub_money(str(raw.get("timeline") or "Scoped together"))[:80],
        "bestFor": _scrub_money(str(raw.get("bestFor") or ""))[:200],
        "includes": includes,
        "badge": badge,
        "highlight": bool(raw.get("highlight")) if plan_id == "growth" else False,
    }


def _normalize_addon(raw: dict[str, Any], idx: int) -> dict[str, Any] | None:
    name = _scrub_money(str(raw.get("name") or "").strip())
    if not name:
        return None
    aid = str(raw.get("id") or f"addon-{idx}").strip().lower()
    aid = re.sub(r"[^a-z0-9\-]+", "-", aid).strip("-") or f"addon-{idx}"
    included = raw.get("includedIn") or []
    if not isinstance(included, list):
        included = []
    included_in = [x for x in included if x in ("launch", "growth")]
    return {
        "id": aid,
        "name": name[:100],
        "description": _scrub_money(str(raw.get("description") or ""))[:280],
        "whyForYou": _scrub_money(str(raw.get("whyForYou") or ""))[:200],
        "includedIn": included_in,
    }


def normalize_build_plans(payload: dict[str, Any]) -> dict[str, Any]:
    plans_raw = {p.get("id"): p for p in (payload.get("plans") or []) if isinstance(p, dict)}
    plans = [_normalize_plan(plans_raw.get(pid) or {"id": pid}, pid) for pid in _PLAN_IDS]
    # Ensure growth is highlighted / badge
    for p in plans:
        if p["id"] == "growth":
            p["highlight"] = True
            if not p.get("badge"):
                p["badge"] = "Most popular"

    addons: list[dict[str, Any]] = []
    seen: set[str] = set()
    for i, raw in enumerate(payload.get("addons") or []):
        if not isinstance(raw, dict):
            continue
        addon = _normalize_addon(raw, i)
        if not addon or addon["id"] in seen:
            continue
        seen.add(addon["id"])
        addons.append(addon)
        if len(addons) >= 8:
            break

    rec = payload.get("recommended_plan_id") or "growth"
    if rec not in _PLAN_IDS:
        rec = "growth"

    return {"recommended_plan_id": rec, "plans": plans, "addons": addons}


def fallback_build_plans(req) -> dict[str, Any]:
    brand = req.business_name or "your business"
    ai = ai_features_from_request(req)
    ai_names = [a.get("name") for a in ai if isinstance(a, dict) and a.get("name")]
    roles = _role_labels(req)
    return normalize_build_plans(
        {
            "recommended_plan_id": "growth",
            "plans": [
                {
                    "id": "launch",
                    "name": "Launch MVP",
                    "tagline": f"Ship the {brand} preview you just saw — for real.",
                    "timeline": "4–8 weeks",
                    "bestFor": "Getting live fast with core customer + owner flows",
                    "includes": [
                        "Production build of your preview (public site + core journeys)",
                        "Owner / admin basics",
                        "AI from your preview, wired for production",
                        "Payments on your main customer path",
                        "Customer confirmations (email + WhatsApp/SMS basics)",
                        "Brand, deploy, and launch handoff",
                    ],
                    "highlight": False,
                },
                {
                    "id": "growth",
                    "name": "Growth MVP",
                    "tagline": "Launch + staff ops, polish, and post-launch care.",
                    "timeline": "8–12 weeks",
                    "bestFor": "Teams that need roles, fuller automations, and care",
                    "includes": [
                        "Everything in Launch MVP",
                        "Staff / role dashboards from your preview",
                        "Richer reminder & follow-up automations",
                        "Domain extras for your industry",
                        "Cinematic polish pass",
                        "30-day care after launch",
                    ],
                    "badge": "Most popular",
                    "highlight": True,
                },
                {
                    "id": "custom",
                    "name": "Custom / Scale",
                    "tagline": "Integrations, multi-location, or ongoing product.",
                    "timeline": "Scoped together",
                    "bestFor": "Complex ops or a longer partnership",
                    "includes": [
                        "Everything in Growth MVP",
                        "Third-party integrations",
                        "Advanced roles & permissions",
                        "Custom workflows and reporting",
                        "Dedicated scope call before we quote",
                    ],
                    "highlight": False,
                },
            ],
            "addons": [
                {
                    "id": "ai-pack",
                    "name": (
                        f"AI suite for {brand}"
                        if len(ai_names) >= 2
                        else f"Production-ready {ai_names[0]}"
                        if ai_names
                        else f"AI assistant for {brand}"
                    ),
                    "description": (
                        f"Wire previewed AI: {', '.join(ai_names)}."
                        if ai_names
                        else "Branded AI helper trained on your offerings and FAQs."
                    ),
                    "whyForYou": "Already in your preview — included in Launch.",
                    "includedIn": ["launch", "growth"],
                },
                {
                    "id": "payments",
                    "name": "Payments on your main journey",
                    "description": "Checkout / deposits wired into the primary action in your preview.",
                    "whyForYou": "Core Launch path — not an upgrade.",
                    "includedIn": ["launch", "growth"],
                },
                {
                    "id": "messaging",
                    "name": "Customer confirmations",
                    "description": "WhatsApp/SMS or email confirmations for the main journey.",
                    "whyForYou": "Included in Launch so customers get real updates.",
                    "includedIn": ["launch", "growth"],
                },
                {
                    "id": "roles",
                    "name": (
                        f"Dashboards for {' + '.join(roles[:3])}"
                        if len(roles) >= 2
                        else "Extra staff / partner roles"
                    ),
                    "description": "Separate live workspaces beyond the owner view.",
                    "whyForYou": "Your preview already defines how the team works.",
                    "includedIn": ["growth"],
                },
                {
                    "id": "care",
                    "name": "30-day care after launch",
                    "description": "Priority fixes and small iterations once you’re live.",
                    "whyForYou": "First month is when real customers find edge cases.",
                    "includedIn": ["growth"],
                },
            ],
        }
    )


def generate_build_plans(
    db: Session,
    request_id: int,
    ai_provider: AIProvider,
    template_renderer: TemplateRenderer,
) -> dict[str, Any]:
    req = get_request(db, request_id)
    if not req.mvp_blueprint and not req.preview_summary:
        raise ValueError("Preview context missing — generate blueprint/preview first.")

    ai_feats = ai_features_from_request(req)
    blueprint_excerpt = (req.mvp_blueprint or "")[:3500]
    prompt = template_renderer.render(
        PromptTemplate.BUILD_PLANS,
        business_name=req.business_name,
        concept_name=req.concept_name or "N/A",
        industry=req.industry or "N/A",
        main_problem=req.main_problem or "N/A",
        desired_outcome=req.desired_outcome or "N/A",
        timeline=req.timeline or "N/A",
        budget_range=req.budget_range or "N/A",
        preview_summary=req.preview_summary or "N/A",
        preview_features=_preview_features(req) or ["(none listed)"],
        ai_features=ai_feats
        or [{"name": "AI helper", "description": "Branded assistant for FAQs and guidance"}],
        role_labels=_role_labels(req) or ["Owner"],
        blueprint_excerpt=blueprint_excerpt or "N/A",
    )

    try:
        raw_text = ai_provider.ask_chat(
            settings.TEXT_MODEL, [{"role": "user", "content": prompt}]
        )
        parsed = extract_json_from_text(raw_text)
        if not isinstance(parsed, dict):
            raise ValueError("Build plans model output was not an object")
        result = normalize_build_plans(parsed)
        if len(result.get("addons") or []) < 3:
            raise ValueError("Too few addons from model")
    except Exception:
        result = fallback_build_plans(req)

    req.build_plans = json.dumps(result)
    req.updated_at = datetime.utcnow()
    db.commit()
    return result


def build_plans_from_request(req) -> dict[str, Any] | None:
    if not getattr(req, "build_plans", None):
        return None
    try:
        data = json.loads(req.build_plans)
        if isinstance(data, dict) and data.get("plans"):
            return normalize_build_plans(data)
    except Exception:
        return None
    return None
